src/server/pps/server/api_server.go

```diff
@@ -3575,3 +3577,105 @@ func newMessageFilterFunc(jqFilter string, projects []*pfs.Project) (func(contex
 		return true, nil
 	}, nil
 }
+
+// This exists in case we end up not ommitting any empty values when marshalling
+// a spec.
+var emptyPipelineSpecJSON string
+
+func init() {
+	var spec pps.PipelineSpec
+	b, err := json.Marshal(spec)
+	if err != nil {
+		panic(fmt.Sprint("could not marshal empty pipeline spec: ", err))
+	}
+	emptyPipelineSpecJSON = string(b)
+}
+
+func (a *apiServer) GetClusterDefaults(ctx context.Context, req *pps.GetClusterDefaultsRequest) (*pps.GetClusterDefaultsResponse, error) {
+	var clusterDefaults pps.ClusterDefaults
+	if err := a.clusterDefaults.ReadOnly(ctx).Get("", &clusterDefaults); err != nil {
+		if !errors.As(err, &col.ErrNotFound{}) {
+			return nil, errors.Wrap(err, "could not read cluster defaults")
+		}
+		clusterDefaults.DetailsJson = "{}"
+	}
+
+	var err error
+	if clusterDefaults.EffectiveDetailsJson, err = jsonMergePatch(emptyPipelineSpecJSON, clusterDefaults.DetailsJson); err != nil {
+		return nil, errors.Wrap(err, "could not merge nil with cluster default details")
+	}
+
+	return &pps.GetClusterDefaultsResponse{ClusterDefaults: &clusterDefaults}, nil
+}
+
+func jsonMergePatch(target, patch string) (string, error) {
+	var targetObject, patchObject any
+	if err := json.Unmarshal([]byte(target), &targetObject); err != nil {
+		return "", errors.Wrap(err, "could not unmarshal target JSON")
+	}
+	if err := json.Unmarshal([]byte(patch), &patchObject); err != nil {
+		return "", errors.Wrap(err, "could not unmarshal patch JSON")
+	}
+	result, err := json.Marshal(mergePatch(targetObject, patchObject))
+	if err != nil {
+		return "", errors.Wrap(err, "could not marshal merge patch result")
+	}
+	return string(result), nil
+}
+
+func mergePatch(target, patch any) any {
+	switch patch := patch.(type) {
+	case map[string]any:
+		var targetMap map[string]any
+		switch t := target.(type) {
+		case map[string]any:
+			targetMap = t
+		default:
+			targetMap = make(map[string]any)
+		}
+		for name, value := range patch {
+			if value == nil {
+				delete(targetMap, name)
+			} else {
+				targetMap[name] = mergePatch(targetMap[name], value)
+			}
+		}
+		return targetMap
+	default:
+		return patch
+	}
+}
+
+func (a *apiServer) SetClusterDefaults(ctx context.Context, req *pps.SetClusterDefaultsRequest) (*pps.SetClusterDefaultsResponse, error) {
+	var d = json.NewDecoder(strings.NewReader(req.GetClusterDefaults().GetDetailsJson()))
+	d.DisallowUnknownFields()
+	if err := d.Decode(&pps.PipelineSpec{}); err != nil {
+		s, err := status.New(codes.InvalidArgument, "invalid pipeline details").WithDetails(&errdetails.BadRequest{
+			FieldViolations: []*errdetails.BadRequest_FieldViolation{
+				{Field: "cluster_defaults.details_json", Description: err.Error()},
+			},
+		})
+		if err != nil {
+			log.Error(ctx, "could not add bad-request details", zap.Error(err))
```

Comment:
I don't like the pattern of error handling error handling.  We should make a StatusWithDetails function that doesn't error.  WithDetails errors if the status is codes.OK; we should just ignore that one.  It additionally errors because it uses the old proto library and calls ptypes.MarshalAny instead of anypb.New.  anypb.New errors when the message is nil or, after MANY levels of indirection, when it can't write to the provided buffer.  The provided buffer is a []byte that it appends to, so it seems like that can never happen.  I'd just restrict the type that can be passed to your WithDetails wrapper, and pass a concrete instance instead of a pointer.  Then no errors can ever occur, and you don't have to write error handling error handling.

Reply:
How do you feel about the new `badRequest` and `unknownError` functions, which always return an error? They do still catch the errors from `WithDetails`, but I think that avoids relying on undocumented internals of other libraries not to change.

Reply:
I'm just saying to abstract this away somehow; the RPC handler doesn't need to know that error-creating functions can return errors.  It just complicates the "meat" of the code with an irrelevant detail that essentially can't happen.  I think restricting the API to 2 types out of the 16 or so error codes is unnecessarily restrictive.  `errors.NewRPC(code, details, proto.Message)`?